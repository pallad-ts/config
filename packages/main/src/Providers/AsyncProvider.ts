import { Provider } from "../Provider";

export abstract class AsyncProvider<T> extends Provider<T> {
    private cache?: Promise<Provider.Value<T>>;

    getValue() {
        if (!this.cache) {
            this.cache = this.resolveValueAsync();
        }
        return this.cache;
    }

    abstract resolveValueAsync(): Promise<Provider.Value<T>>;
}
