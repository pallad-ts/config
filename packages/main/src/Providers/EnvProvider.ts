import {Provider} from "../Provider";

export class EnvProvider<T> extends Provider<string> {
    constructor(private key: string,
                private envs: typeof process['env'] = process.env) {
        super();
    }

    getDescription(): string {
        return 'ENV: ' + this.key;
    }

    isAvailable() {
        if (this.key in this.envs) {
            return this.envs[this.key] !== '';
        }
        return false;
    }

    protected retrieveValue() {
        return this.envs[this.key]!;
    }
}
