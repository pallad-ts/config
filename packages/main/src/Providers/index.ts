export * from './DefaultValueProvider';
export * from './EnvProvider';
export * from './FirstAvailableProvider';
export * from './FormattedStringProvider';
export * from './TransformProvider';
export * from './wrapWithDefaultAndTransformer';
