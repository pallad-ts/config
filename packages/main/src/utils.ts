import {Provider} from './Provider';

export type OptionalPromise<T> = T | Promise<T>;
export type UnwrapProvider<T> = T extends Provider<infer U> ? U : T;
