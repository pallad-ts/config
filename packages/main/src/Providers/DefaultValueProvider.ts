import {Provider} from "../Provider";
import {OptionalPromise} from '../utils';
import {isPromise} from '../common/isPromise';
import {runOnOptionalPromise} from '../common/runOnOptionalPromise';

/**
 * Wrapper for other provider. Returns default value if given provider has no available value.
 *
 * @public
 */
export class DefaultValueProvider<T, TOriginal> extends Provider<T | TOriginal> {
    constructor(private provider: Provider<TOriginal>, private defaultValue: T) {
        super();
    }

    getDescription(): string {
        return `${this.provider.getDescription()} - default: ${this.defaultValue}`;
    }

    isAvailable() {
        return true;
    }

    protected retrieveValue(): OptionalPromise<T | TOriginal> {
        return runOnOptionalPromise(
            this.provider.isAvailable(),
            isAvailable => {
                if (isAvailable) {
                    return this.provider.getValue() as any;
                }
                return this.defaultValue;
            }
        );
    }

    static optionalWrap<T, TOriginal>(provider: Provider<TOriginal>, defaultValue?: T) {
        if (defaultValue !== undefined) {
            return new DefaultValueProvider<T, TOriginal>(provider, defaultValue);
        }
        return provider;
    }
}
