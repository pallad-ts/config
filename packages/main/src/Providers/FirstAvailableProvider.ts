import {Provider} from "../Provider";
import {isPromise} from '../common/isPromise';
import {OptionalPromise} from '../utils';
import {runOnOptionalPromise} from '../common/runOnOptionalPromise';

export class FirstAvailableProvider<T extends Array<Provider<any>>> extends Provider<FirstAvailableProvider.Unwrap<T>> {
    private providers: Array<Provider<T>>;

    constructor(...providers: T) {
        super();
        this.providers = providers;
    }

    isAvailable() {
        return runOnOptionalPromise(this.getFirstAvailableProvider(), provider => !!provider);
    }

    private getFirstAvailableProvider(): OptionalPromise<Provider<any> | undefined> {
        const iterator = this.providers[Symbol.iterator]();
        for (const provider of iterator) {
            const isAvailable = provider.isAvailable();
            if (isPromise(isAvailable)) {
                return isAvailable.then(x => {
                    if (x) {
                        return provider;
                    }
                    return (async () => {
                        for (const provider of iterator) {
                            const isAvailable = await provider.isAvailable();
                            if (isAvailable) {
                                return provider;
                            }
                        }
                    })();
                });
            } else if (isAvailable) {
                return provider;
            }
        }
    }

    getDescription(): string {
        const desc = this.providers.map(x => x.getDescription()).join(', ');
        return `First available - ${desc}`;
    }

    retrieveValue() {
        return runOnOptionalPromise(
            this.getFirstAvailableProvider(),
            provider => {
                return provider!.getValue()
            }
        );
    }
}

export namespace FirstAvailableProvider {
    export type Unwrap<T> = T extends Array<Provider<infer U>> ? U : T;
}
