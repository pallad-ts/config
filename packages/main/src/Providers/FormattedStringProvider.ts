import {Provider} from "../Provider";
import {OptionalPromise} from '../utils';
import {isPromise} from '../common/isPromise';

function produce(strings: readonly string[], values: any[]) {
    let str = '';
    for (const [index, string] of strings.entries()) {
        str += `${string}${(values[index] || '')}`;
    }
    return str;
}

export class FormattedStringProvider extends Provider<string> {
    constructor(private strings: readonly string[], private values: any[]) {
        super();
    }

    getDescription(): string {
        const desc = produce(this.strings, this.values.map(x => Provider.is(x) ? `\${${x.getDescription()}}` : x));
        return `Formatted string ("${desc}")`;
    }

    isAvailable(): OptionalPromise<boolean> {
        const providersAvailability = this.getDependentProviders().map(x => x.isAvailable());
        const hasPromises = providersAvailability.some(isPromise);
        if (hasPromises) {
            return Promise.all(providersAvailability)
                .then(x => {
                    return x.every(x => x);
                })
        }
        return providersAvailability.every(x => x);
    }

    getDependentProviders(): Array<Provider<any>> {
        return this.values.filter(Provider.is);
    }

    retrieveValue(): OptionalPromise<string> {
        const values = this.values.map(x => Provider.is(x) ? x.getValue() : x)

        const hasPromises = values.some(isPromise);
        if (hasPromises) {
            return Promise.all(values)
                .then(x => {
                    return produce(this.strings, x);
                })
        }
        return produce(this.strings, values);
    }

    static create(strings: readonly string[], ...values: any[]) {
        if (values.some(Provider.is)) {
            return new FormattedStringProvider(strings, values);
        }
        return produce(strings, values);
    }
}
