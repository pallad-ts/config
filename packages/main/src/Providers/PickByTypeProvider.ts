import {Provider} from '../Provider';
import {extractProvidersFromConfig} from '../common/extractProvidersFromConfig';
import {Config} from '../Config';
import {runOnOptionalPromise} from '../common/runOnOptionalPromise';
import {ERRORS} from '../errors';
import {isPromise} from '../common/isPromise';
import {loadConfig} from '../common/loadConfig';
import {OptionalPromise} from '../utils';

export class PickByTypeProvider<T> extends Provider<T> {
    private options = new Map<string, any>();

    private constructor(private typeProvider: Provider<string>) {
        super();
    }

    static create(typeDependency: Provider<string>) {
        return new PickByTypeProvider<undefined>(typeDependency);
    }

    registerOptions<TType extends string, TValue>(
        type: TType,
        value: TValue
    ): PickByTypeProvider<NonNullable<T> | PickByTypeProvider.Value<TType, Config.ResolvedValue<TValue>>> {
        if (this.options.has(type)) {
            throw ERRORS.PICK_PROVIDER_TYPE_ALREADY_EXISTS.format(type);
        }
        this.options.set(type, value);
        return this as any;
    }

    private getType(): OptionalPromise<string> {
        return runOnOptionalPromise(
            this.typeProvider.getValue(),
            resolvedType => {
                if (!this.options.has(resolvedType)) {
                    throw ERRORS.PICK_PROVIDER_UNREGISTERED_TYPE.format(
                        resolvedType,
                        Array.from(this.options.keys()).join(', ')
                    );
                }
                return resolvedType;
            }
        );
    }

    private getOptions() {
        return runOnOptionalPromise(this.getType(), type => this.options.get(type)!);
    }

    protected retrieveValue() {
        return runOnOptionalPromise(this.getType(), type => {
            return runOnOptionalPromise(this.getOptions(), options => {
                const finalOptions = Provider.is(options) ? options.getValue() : loadConfig(options);
                return runOnOptionalPromise(finalOptions, result => {
                    return {
                        type,
                        options: result
                    } as any as T;
                })
            });
        });
    }

    getDescription() {
        return `Pick by type from ${this.typeProvider.getDescription()}`;
    }

    private getDependentProviders() {
        return runOnOptionalPromise(
            this.getOptions(),
            options => {
                const optionsDependencies = Provider.is(options) ? [options] : extractProvidersFromConfig(options);
                return [
                    this.typeProvider,
                    ...optionsDependencies
                ];
            }
        );
    }

    isAvailable(): OptionalPromise<boolean> {
        return runOnOptionalPromise<boolean, OptionalPromise<boolean>>(
            this.typeProvider.isAvailable(),
            isTypeAvailable => {
                if (!isTypeAvailable) {
                    return false;
                }

                return runOnOptionalPromise(
                    this.getDependentProviders(),
                    dependentProviders => {
                        const iterator = dependentProviders[Symbol.iterator]();
                        for (const provider of iterator) {
                            const isAvailable = provider.isAvailable();
                            if (isPromise(isAvailable)) {
                                return isAvailable.then(result => {
                                    if (!result) {
                                        return false;
                                    }

                                    return (async () => {
                                        for (const provider of iterator) {
                                            const isAvailable = await provider.isAvailable();
                                            if (!isAvailable) {
                                                return false;
                                            }
                                        }
                                        return true;
                                    })();
                                })
                            } else if (!isAvailable) {
                                return false;
                            }
                        }
                        return true;
                    }
                )
            }
        );
    }
}

export namespace PickByTypeProvider {
    export interface Value<TType extends string, TOptions> {
        type: TType;
        options: TOptions;
    }
}
