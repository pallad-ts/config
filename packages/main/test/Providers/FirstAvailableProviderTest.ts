import {assert, IsExact} from 'conditional-type-checks';
import {FirstAvailableProvider} from '@src/Providers';
import {DummyProvider} from '../dummies/DummyProvider';
import {OptionalPromise} from '@src/utils';

describe('FirstAvailableProvider', () => {

    it('types', () => {
        const provider = new FirstAvailableProvider(
            DummyProvider.sync<'foo'>({value: 'foo', isAvailable: true}),
            DummyProvider.sync<'bar'>({value: 'bar', isAvailable: true}),
        );
        const value = provider.getValue();
        assert<IsExact<typeof value, OptionalPromise<'foo' | 'bar'>>>(true);
    });

    it('description', () => {
        const provider = new FirstAvailableProvider(
            DummyProvider.sync({value: 'foo', isAvailable: true, description: 'desc1'}),
            DummyProvider.sync({value: 'foo2', isAvailable: true, description: 'desc2'}),
            DummyProvider.sync({value: 'foo3', isAvailable: true, description: 'desc3'}),
        );

        expect(provider.getDescription())
            .toMatchSnapshot();
    });

    describe('sync', () => {
        describe('getting value', () => {
            it('first one is available', () => {
                const provider = new FirstAvailableProvider(
                    DummyProvider.sync({value: 'foo', isAvailable: true}),
                    DummyProvider.sync({value: 'bar', isAvailable: true}),
                );
                expect(provider.getValue())
                    .toEqual('foo');
            });

            it('second one is available', () => {
                const provider = new FirstAvailableProvider(
                    DummyProvider.sync({value: 'foo', isAvailable: false}),
                    DummyProvider.sync({value: 'bar', isAvailable: true}),
                );
                expect(provider.getValue())
                    .toEqual('bar');
            });

            it('throw an error if none is available', () => {
                const provider = new FirstAvailableProvider(
                    DummyProvider.sync({value: 'foo', isAvailable: false}),
                    DummyProvider.sync({value: 'bar', isAvailable: false}),
                );

                expect(() => {
                    provider.getValue()
                })
                    .toThrowErrorMatchingSnapshot();
            });
        });

        describe('availability', () => {
            it('available if any of providers is available', () => {
                const provider = new FirstAvailableProvider(
                    DummyProvider.sync({value: 'foo', isAvailable: false}),
                    DummyProvider.sync({value: 'bar', isAvailable: false}),
                    DummyProvider.sync({value: 'baz', isAvailable: true}),
                );

                expect(provider.isAvailable())
                    .toBe(true);
            });

            it('not available if none of providers is available', () => {
                const provider = new FirstAvailableProvider(
                    DummyProvider.sync({value: 'foo', isAvailable: false}),
                    DummyProvider.sync({value: 'bar', isAvailable: false}),
                    DummyProvider.sync({value: 'baz', isAvailable: false}),
                );

                expect(provider.isAvailable())
                    .toBe(false);
            });
        });
    });

    describe('async', () => {
        describe('getting value', () => {
            it('first one is available', () => {
                const provider = new FirstAvailableProvider(
                    DummyProvider.async({value: 'foo', isAvailable: true}),
                    DummyProvider.async({value: 'bar', isAvailable: true}),
                );
                return expect(provider.getValue())
                    .resolves
                    .toEqual('foo');
            });

            it('second one is available', () => {
                const provider = new FirstAvailableProvider(
                    DummyProvider.async({value: 'foo', isAvailable: false}),
                    DummyProvider.async({value: 'bar', isAvailable: true}),
                );
                return expect(provider.getValue())
                    .resolves
                    .toEqual('bar');
            });

            it('throw an error if none is available', () => {
                const provider = new FirstAvailableProvider(
                    DummyProvider.async({value: 'foo', isAvailable: false}),
                    DummyProvider.async({value: 'bar', isAvailable: false}),
                );

                return expect(provider.getValue())
                    .rejects
                    .toThrowErrorMatchingSnapshot();
            });
        });

        describe('availability', () => {
            it('available if any of providers is available', () => {
                const provider = new FirstAvailableProvider(
                    DummyProvider.async({value: 'foo', isAvailable: false}),
                    DummyProvider.async({value: 'bar', isAvailable: false}),
                    DummyProvider.async({value: 'baz', isAvailable: true}),
                );

                return expect(provider.isAvailable())
                    .resolves
                    .toBe(true);
            });

            it('not available if none of providers is available', () => {
                const provider = new FirstAvailableProvider(
                    DummyProvider.async({value: 'foo', isAvailable: false}),
                    DummyProvider.async({value: 'bar', isAvailable: false}),
                    DummyProvider.async({value: 'baz', isAvailable: false}),
                );

                return expect(provider.isAvailable())
                    .resolves
                    .toBe(false);
            });
        });
    });

    describe('sync + async', () => {
        describe('getting value', () => {
            it('first one is available', () => {
                const provider = new FirstAvailableProvider(
                    DummyProvider.sync({value: 'foo', isAvailable: false}),
                    DummyProvider.async({value: 'bar', isAvailable: false}),
                    DummyProvider.async({value: 'bar2', isAvailable: false}),
                    DummyProvider.async({value: 'baz', isAvailable: true}),
                );
                return expect(provider.getValue())
                    .resolves
                    .toEqual('baz');
            });
        });
    });
});
