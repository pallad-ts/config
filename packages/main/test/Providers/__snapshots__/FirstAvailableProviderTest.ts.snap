// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FirstAvailableProvider async getting value throw an error if none is available 1`] = `"Config value not available: First available - DummyProvider, DummyProvider"`;

exports[`FirstAvailableProvider description 1`] = `"First available - desc1, desc2, desc3"`;

exports[`FirstAvailableProvider sync getting value throw an error if none is available 1`] = `"Config value not available: First available - DummyProvider, DummyProvider"`;
